coverage_analysis/functions_analysed/strassen_coverage.py

```diff
@@ -0,0 +1,162 @@
+    e, f, g, h = split_matrix(matrix_b)
+
+    t1 = actual_strassen(a, matrix_subtraction(f, h))
+    t2 = actual_strassen(matrix_addition(a, b), h)
+    t3 = actual_strassen(matrix_addition(c, d), e)
+    t4 = actual_strassen(d, matrix_subtraction(g, e))
+    t5 = actual_strassen(matrix_addition(a, d), matrix_addition(e, h))
+    t6 = actual_strassen(matrix_subtraction(b, d), matrix_addition(g, h))
+    t7 = actual_strassen(matrix_subtraction(a, c), matrix_addition(e, f))
+
+    top_left = matrix_addition(matrix_subtraction(matrix_addition(t5, t4), t2), t6)
+    top_right = matrix_addition(t1, t2)
+    bot_left = matrix_addition(t3, t4)
+    bot_right = matrix_subtraction(matrix_subtraction(matrix_addition(t1, t5), t3), t7)
+
+    # construct the new matrix from our 4 quadrants
+    new_matrix = []
+    for i in range(len(top_right)):
+        new_matrix.append(top_left[i] + top_right[i])
+    for i in range(len(bot_right)):
+        new_matrix.append(bot_left[i] + bot_right[i])
+    return new_matrix
+
+def matrix_addition(matrix_a: list, matrix_b: list):
+    return [
+        [matrix_a[row][col] + matrix_b[row][col] for col in range(len(matrix_a[row]))]
+        for row in range(len(matrix_a))
+    ]
+
+
+def matrix_subtraction(matrix_a: list, matrix_b: list):
+    return [
+        [matrix_a[row][col] - matrix_b[row][col] for col in range(len(matrix_a[row]))]
+        for row in range(len(matrix_a))
+    ]
+
+def split_matrix(a: list) -> tuple[list, list, list, list]:
+    """
+    Given an even length matrix, returns the top_left, top_right, bot_left, bot_right
+    quadrant.
+
+    >>> split_matrix([[4,3,2,4],[2,3,1,1],[6,5,4,3],[8,4,1,6]])
+    ([[4, 3], [2, 3]], [[2, 4], [1, 1]], [[6, 5], [8, 4]], [[4, 3], [1, 6]])
+    >>> split_matrix([
+    ...     [4,3,2,4,4,3,2,4],[2,3,1,1,2,3,1,1],[6,5,4,3,6,5,4,3],[8,4,1,6,8,4,1,6],
+    ...     [4,3,2,4,4,3,2,4],[2,3,1,1,2,3,1,1],[6,5,4,3,6,5,4,3],[8,4,1,6,8,4,1,6]
+    ... ])  # doctest: +NORMALIZE_WHITESPACE
+    ([[4, 3, 2, 4], [2, 3, 1, 1], [6, 5, 4, 3], [8, 4, 1, 6]], [[4, 3, 2, 4],
+      [2, 3, 1, 1], [6, 5, 4, 3], [8, 4, 1, 6]], [[4, 3, 2, 4], [2, 3, 1, 1],
+      [6, 5, 4, 3], [8, 4, 1, 6]], [[4, 3, 2, 4], [2, 3, 1, 1], [6, 5, 4, 3],
+      [8, 4, 1, 6]])
+    """
+    if len(a) % 2 != 0 or len(a[0]) % 2 != 0:
+        raise Exception("Odd matrices are not supported!")
+
+    matrix_length = len(a)
+    mid = matrix_length // 2
+
+    top_right = [[a[i][j] for j in range(mid, matrix_length)] for i in range(mid)]
+    bot_right = [
+        [a[i][j] for j in range(mid, matrix_length)] for i in range(mid, matrix_length)
+    ]
+
+    top_left = [[a[i][j] for j in range(mid)] for i in range(mid)]
+    bot_left = [[a[i][j] for j in range(mid)] for i in range(mid, matrix_length)]
+
+    return top_left, top_right, bot_left, bot_right
+
+def strassen(matrix1: list, matrix2: list, coverage: CoverageData) -> list:
+    if matrix_dimensions(matrix1)[1] != matrix_dimensions(matrix2)[0]:
```

Comment:
check the requirement 3 under 3.5.1, each time you've a "if" statement, if the "else" is missing, you need to add it. i.e you can just add:
`else:
    coverage.log_branch(type, x)
`

Reply:
Sure, I'll add that.